### ISSUE TITLE
Compression support?

### ISSUE THREAD

**Issue by sezaru** [lucaong/cubdb#20]
Do you think we can somehow support some kind of data compression?

My guess is that compress the whole database is not possible since it would pretty much break efficient query and the append logic CubDB uses.

But I was wondering if we can add an option to compress the value part of a key/value when CubDB uses `:erlang.term_to_binary()` by piping it to `:zlib.gzip()`.

Of course that if the value being stored is small, running `:zlib.gzip()` will actually increase the final byte size of the data stored, so this option would probably be turned off by default.

The advantage comes when you want to store huge data as the value, this can make the value stored way smaller than default (which is one of the use cases I want to use CubDB with).

I can try creating a PR if you like the idea.

Best regards, Eduardo.

**Comment by sezaru** [comment]
So, I was looking a little deeper into it, and I realized that you created a protocol called `CubDB.Store` which you then implemented it using `CubDB.Store.File`.

As far as I understood, you create it during [this line](https://github.com/lucaong/cubdb/blob/master/lib/cubdb.ex#L813).

So, what about instead of adding compression support hardcoded to CubDB, we instead make the `CubDB.Store` a parameter for the `CubDB` gen_server with the `CubDB.Store.File` as the default one.

That way people can develop new stores for CubDB using other solutions like snappy, memory, etc.

**Comment by lucaong** [comment]
Hi @sezaru ,
Sorry for the late response. I think that compression could be a nice feature, but it’s not trivial to implement properly. In fact, in an earlier version, CubDB used to support compression using the option provided by the Erlang `term_to_binary` function. Unfortunately, that proved to negatively affect performance quite a lot (especially for large values, that would benefit more for compression), so I removed it before the release candidates. Using some other compression algorithm such as Snappy might yield better results, but it is something that requires thought and planning. For now, I decided to leave compression to the user (one can compress values before writing them in user code).

Letting the user specify their own Store could be a way to go, but for now properly implementing a store is not necessarily trivial. I definitely want to put some thoughts in that, not only for compression but also for caching: both are features that have trade-offs and are not always desirable. I am looking for a good API to give control on them to the user. I just think that getting the API wrong would be a bigger problem than delaying those features. After all, both caching and compression can be implemented at a higher level in user code.
